ydb/core/kqp/provider/rewrite_io_utils.cpp

```diff
@@ -0,0 +1,169 @@
+    childrenToChange.emplace_back(newChild);
+    parent->ChangeChildrenInplace(std::move(childrenToChange));
+}
+
+void ChangeChild(const TExprNode::TPtr& parent, ui32 index, const TExprNode::TPtr& newChild) {
+    YQL_ENSURE(parent->ChildrenSize() > index);
+
+    auto childrenToChange = parent->ChildrenList();
+    childrenToChange[index] = newChild;
+    parent->ChangeChildrenInplace(std::move(childrenToChange));
+}
+
+TExprNode::TPtr FindSavedQueryGraph(const TExprNode::TPtr& carrier) {
+    if (carrier->ChildrenSize() == 0) {
+        return nullptr;
+    }
+    auto lastChild = carrier->Children().back();
+    return lastChild->IsCallable(QueryGraphNodeSignature) ? lastChild->ChildPtr(0) : TExprNode::TPtr();
+}
+
+void SaveQueryGraph(const TExprNode::TPtr& carrier, TExprContext& ctx, const TExprNode::TPtr& payload) {
+    AddChild(carrier, ctx.NewCallable(payload->Pos(), QueryGraphNodeSignature, {payload}));
+}
+
+void InsertExecutionOrderDependencies(const TExprNode::TPtr& queryGraph, const TExprNode::TPtr& worldBefore) {
+    VisitExpr(
+        queryGraph,
+        nullptr,
+        [&worldBefore](const TExprNode::TPtr& node) {
+            if (node->ChildrenSize() > 0 && node->Child(0)->IsWorld()) {
+                ChangeChild(node, 0, worldBefore);
+            }
+            return true;
+        }
+    );
+}
+
+bool CheckTopLevelness(const TExprNode::TPtr& candidateRead, const TExprNode::TPtr& queryGraph) {
+    THashSet<TExprNode::TPtr> readsInCandidateSubgraph;
+    VisitExpr(candidateRead, [&readsInCandidateSubgraph](const TExprNode::TPtr& node) {
+        if (node->IsCallable(ReadName)) {
+            readsInCandidateSubgraph.emplace(node);
+        }
+        return true;
+    });
+
+    return !FindNode(queryGraph, [&readsInCandidateSubgraph](const TExprNode::TPtr& node) {
+        return node->IsCallable(ReadName) && !readsInCandidateSubgraph.contains(node);
+    });
+}
+
+TExprNode::TPtr FindTopLevelRead(const TExprNode::TPtr& queryGraph) {
+    const TExprNode::TPtr* lastReadInTopologicalOrder = nullptr;
+    VisitExpr(
+        queryGraph,
+        nullptr,
+        [&lastReadInTopologicalOrder](const TExprNode::TPtr& node) {
+            if (node->IsCallable(ReadName)) {
+                lastReadInTopologicalOrder = &node;
+            }
+            return true;
+        }
+    );
+
+    if (!lastReadInTopologicalOrder) {
+        return nullptr;
+    }
+
+    YQL_ENSURE(CheckTopLevelness(*lastReadInTopologicalOrder, queryGraph),
+               "Info for developers: assumption that there is only one top level Read! is wrong "
+               "for the expression graph of the query stored in the view:\n"
+                   << queryGraph->Dump());
+
+    return *lastReadInTopologicalOrder;
+}
+
+}
+
+TExprNode::TPtr RewriteReadFromView(
+    const TExprNode::TPtr& node,
+    TExprContext& ctx,
+    const TString& query,
+    const TString& cluster
+) {
+    const auto readNode = node->ChildPtr(0);
```

Comment:
What kind of node we expect here? Use typed nodes or at least add comment.

Reply:
[Used](https://github.com/jepett0/ydb/blob/586058fd4169e192e0902fd05db5e933c0a21c69/ydb/core/kqp/provider/rewrite_io_utils.cpp#L135) TCoRead to express my intention better